### PR TITLE
Extending ML problems

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

One-dimension ML is really boring.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/benchmark/experiments.py [hunk 1]
```diff
@@ -42,7 +42,9 @@ def mltuning(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
     # First, a few functions with constraints.
     functions = [
         MLTuning("1d_decision_tree_regression"),
-        MLTuning("1d_decision_tree_regression_full")
+        MLTuning("1d_decision_tree_regression_full"),
+        MLTuning("2d_decision_tree_regression_full"),
+        MLTuning("3d_decision_tree_regression_full"),
```
Comment (jrapin) [inline nevergrad/benchmark/experiments.py:47]: instead of having so many different names, maybe you would need "dimension" and "full" parameters instead?
And for test cases which are very different, creating a different class may be simpler
I'm a bit afraid there starts to be a lot of `if` conditions everywhere, this is never a good sign

Comment (teytaud) [reply nevergrad/benchmark/experiments.py:47]: I'm not sure I agree... we will have other test cases, including sophisticated cases in which Nevergrad will choose between various MLP and Decision trees and others. So we will need a function with everything.

Comment (teytaud) [reply nevergrad/benchmark/experiments.py]: Thanks for the offline comments, which help a lot :-)   I update accordingly.

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 2]
```diff
@@ -25,59 +26,116 @@ class MLTuning(ExperimentFunction):
...
             result += np.sum((y_test - pred_test)**2)
         return result / num_data
 
-    def __init__(self, problem_type: str):
-        self.problem_type = problem_type
+    def __init__(self, regressor: str, dimension: int):
+        self.regressor = regressor
 
-        if problem_type == "1d_decision_tree_regression":
+        if regressor == "decision_tree_depth_regression":
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:82]: why not simply `decision_tree`?

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 3]
```diff
@@ -25,59 +26,116 @@ class MLTuning(ExperimentFunction):
...
             assert isinstance(depth, int), f"depth has class {type(depth)} and value {depth}."
     
             # Fit regression model
-            regr = DecisionTreeRegressor(max_depth=depth, criterion=criterion,
-                                         min_samples_split=min_samples_split)
+            if regressor == "decision_tree":
+                regr = DecisionTreeRegressor(max_depth=depth, criterion=criterion,
+                                             min_samples_split=min_samples_split)
+            elif regressor == "mlp":
+                regr = MLPRegressor(alpha=alpha, activation=activation, solver=solver, learning_rate=learning_rate)
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:71]: else raise an error?

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 4]
```diff
@@ -25,59 +26,116 @@ class MLTuning(ExperimentFunction):
...
+            )        
+            super().__init__(partial(self._ml_parametrization, dimension=dimension, noise_free=False,        
+                                     alpha=1.0, learning_rate="no", regressor="decision_tree", 
+                                     activation="no", solver="no"), parametrization)
+            self.evaluation_function = partial(self._ml_parametrization, dimension=dimension, criterion="mse",  # type: ignore
+                                               min_samples_split=0.00001,
+                                               regressor="decision_tree", noise_free=True,        
+                                               alpha=1.0, learning_rate="no", 
+                                               activation="no", solver="no")
+        elif regressor == "nn":
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:126]: why not `mlp` if that's what you provide to `regressor`? that would make it more uniform

File: nevergrad/functions/ml/test_mlfunctionlib.py [hunk 5]
```diff
@@ -12,12 +12,15 @@
 
 
 def test_ml_tuning() -> None:
-    func = mlfunctionlib.MLTuning("1d_decision_tree_regression")
+    func = mlfunctionlib.MLTuning("decision_tree_depth_regression", 1)
     x: int = 3
     y1 = func(x)  # returns a float
     y2 = func(x)  # returns the same float
     np.testing.assert_array_almost_equal(y1, y2)
     y3 = func.evaluation_function(x)   # returns a float
```
Comment (jrapin) [inline nevergrad/functions/ml/test_mlfunctionlib.py:20]: testing agains actual value would help check we do not modify the function without noticing. One test for each case with pytest framework would be perfect.
Eg:
https://github.com/facebookresearch/nevergrad/blob/099673586a544fea891c88209fd517875b56dbe2/nevergrad/functions/photonics/test_core.py#L117-L123

Comment (teytaud) [reply nevergrad/functions/ml/test_mlfunctionlib.py:20]: Done, thx.
This also helped the code itself: the training had to be made deterministic by setting e.g. random_state=0.
Not sure how much this is a desirable feature btw.

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/functions/ml/mlfunctionlib.py]: Thx; I've "asserted" for the last case.

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/functions/ml/mlfunctionlib.py]: Right. I switch to mlp everywhere.

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/ml/mlfunctionlib.py]: Yes, I remove the useless suffix.

File: nevergrad/functions/ml/test_mlfunctionlib.py (hunk 5 above)
Comment (jrapin) [reply nevergrad/functions/ml/test_mlfunctionlib.py:20]: probably not desirable at all if you want to have variability

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 6]
```diff
@@ -25,59 +26,119 @@ class MLTuning(ExperimentFunction):
...
+                            regressor: str,  # Choice of learner.
+                            noise_free: bool):
         # 10-folds cross-validation
-        num_data: int = 80
+        num_data: int = 120
         result: float = 0.
         for cv in range(10):
             # All data.
             X_all = np.arange(0., 1., 1. / num_data)
+            random_state = np.random.RandomState(17)
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:46]: not sure it's a good idea, maybe providing a seed to the function which is pulled from the random state in the parametrization would be better. This way the function is random and the randomness can be controlled easily through the parametrization
